File: sdks/go/connect_client.go
package client

import (
	"context"
	"errors"
	"fmt"
	"io"
	"iter"
	"net/http"
	"strings"

	"connectrpc.com/connect"
	"google.golang.org/grpc"

	pb "github.com/anaregdesign/lantern/pb/graph/v1"
	"github.com/anaregdesign/lantern/pb/graph/v1/graphv1connect"
)

// ConnectOption configures the additive Connect-backed Lantern constructor.
// It is intentionally NOT compatible with the existing Option type — the
// underlying transports take fundamentally different knobs (http.Client vs
// grpc.DialOption), and overloading one shape across two transports would
// produce confusing runtime behavior at every WithXxx site.
type ConnectOption func(*connectOptions)

type connectOptions struct {
	httpClient     *http.Client
	clientOptions  []connect.ClientOption
	defaultTimeout int64 // nanoseconds; stored as int64 to skip the time import.
	batchChunkSize int
}

// WithHTTPClient supplies the http.Client used by the Connect transport.
// When omitted, NewLanternConnect builds an h2c-capable client via
// defaultH2CClient (HTTP/2 over plaintext) so the SDK works out of the box
// against the server's additive Connect listener.
//
// Pass a TLS-configured http.Client for production: build an
// http2.Transport with a real *tls.Config and supply it via
// &http.Client{Transport: ...}. The Connect protocol does NOT require h2c —
// HTTPS is the recommended production transport.
func WithHTTPClient(c *http.Client) ConnectOption {
	return func(o *connectOptions) { o.httpClient = c }
}

// WithConnectClientOption forwards arbitrary connect.ClientOption values
// (interceptors, codec selection, gzip compression, etc.) to the generated
// Connect client constructor. Callers needing the gRPC wire protocol pass
// connect.WithGRPC() here; callers wanting OTel pass an OTel interceptor.
func WithConnectClientOption(opts ...connect.ClientOption) ConnectOption {
	return func(o *connectOptions) { o.clientOptions = append(o.clientOptions, opts...) }
}

// WithConnectBatchChunkSize mirrors WithBatchChunkSize for the Connect-
// backed client. Must be > 0; otherwise the default of 1000 is kept.
func WithConnectBatchChunkSize(n int) ConnectOption {
	return func(o *connectOptions) {
		if n > 0 {
			o.batchChunkSize = n
		}
	}
}

// NewLanternConnect returns a Lantern client backed by the Connect-Go
// transport. baseURL must include the scheme: `http://host:port` for h2c
// (matches LANTERN_CONNECT_PORT default), or `https://host` for TLS.
//
// This constructor is the **additive** preview of the future v1.0 transport
// (see #335, #337, #338). It coexists with the existing grpc-backed
// NewLantern / NewLanternWithEndpoints constructors; both return the same
// *Lantern type, so downstream code that holds a *Lantern is transport-
// agnostic. Pick the constructor that matches your wire-side reality:
//
//   - NewLanternConnect("http://lantern:6381") — server with
//     LANTERN_CONNECT_PORT=6381 (additive Connect listener from #337).
//   - NewLantern("lantern:6380") — server with the primary gRPC port
//     (default until #347 cuts over).
//
// baseURL must not end with a trailing slash; the Connect-Go client
// constructs paths by concatenation and a trailing slash would produce
// `//graph.v1.LanternService/...` URLs that the server rejects. A trailing
// slash supplied by the caller is stripped defensively.
func NewLanternConnect(baseURL string, opts ...ConnectOption) (*Lantern, error) {
	if baseURL == "" {
		return nil, errors.New("client: NewLanternConnect requires a base URL with scheme")
	}
	if !strings.HasPrefix(baseURL, "http://") && !strings.HasPrefix(baseURL, "https://") {
		return nil, fmt.Errorf("client: NewLanternConnect baseURL must start with http:// or https://; got %q", baseURL)
	}
	baseURL = strings.TrimRight(baseURL, "/")

	co := connectOptions{
		batchChunkSize: defaultBatchChunkSize,
	}
	for _, apply := range opts {
		apply(&co)
	}
	if co.httpClient == nil {
		co.httpClient = defaultH2CClient()
	}

	cc := graphv1connect.NewLanternServiceClient(co.httpClient, baseURL, co.clientOptions...)
	return &Lantern{
		conn:   nil, // No grpc.ClientConn on this transport path.
		client: &connectClientAdapter{svc: cc},
		opts: options{
			batchChunkSize:    co.batchChunkSize,
			serviceConfigJSON: "", // unused by Connect transport
		},
		connectHTTPClient: co.httpClient,
		connectBaseURL:    baseURL,
	}, nil
}

// connectClientAdapter implements pb.LanternServiceClient on top of the
// generated Connect-Go client. This is the single seam that lets every
// existing *Lantern method (GetVertex, PutVertices, AddEdges, Illuminate,
// ...) keep its grpc-style signature while the underlying transport is
// Connect.
//
// The methods accept `...grpc.CallOption` to satisfy the generated
// interface but IGNORE them — Connect-Go has its own per-call options
// (request headers, deadlines via context) that callers configure via
// WithConnectClientOption at construction time. *Lantern's own helpers
// never pass call options to these methods, so the ignored variadic is
// invisible to consumers; direct users of the underlying pb client
// interface should construct the Connect client directly.
type connectClientAdapter struct {
	svc graphv1connect.LanternServiceClient
}

// connectUnary is the one-line forwarding helper every adapter method
// uses. Generics keep each method body to a single line while preserving
// type safety: the call site supplies the typed underlying Connect-Go
// method, and the wrapper handles request/response boxing plus error
// translation.
func connectUnary[Req, Resp any](
	ctx context.Context,
	req *Req,
	fn func(context.Context, *connect.Request[Req]) (*connect.Response[Resp], error),
) (*Resp, error) {
	resp, err := fn(ctx, connect.NewRequest(req))
	if err != nil {
		return nil, connectErrToGRPC(err)
	}
	return resp.Msg, nil
}

// connectErrToGRPC repackages a Connect error into a gRPC status error
// carrying the equivalent code, so the existing wrapStatus shim in
// client.go (and any consumer's `status.Code(err)` call) keep returning
// the same canonical codes after the transport switch.
//
// Defined in connect_errors.go so the code-mapping table sits with the
// rest of the gRPC interop bridge.

func (a *connectClientAdapter) Illuminate(ctx context.Context, req *pb.IlluminateRequest, _ ...grpc.CallOption) (*pb.IlluminateResponse, error) {
	return connectUnary(ctx, req, a.svc.Illuminate)
}
func (a *connectClientAdapter) GetVertex(ctx context.Context, req *pb.GetVertexRequest, _ ...grpc.CallOption) (*pb.GetVertexResponse, error) {
	return connectUnary(ctx, req, a.svc.GetVertex)
}
func (a *connectClientAdapter) GetVertices(ctx context.Context, req *pb.GetVerticesRequest, _ ...grpc.CallOption) (*pb.GetVerticesResponse, error) {
	return connectUnary(ctx, req, a.svc.GetVertices)
}
func (a *connectClientAdapter) PutVertex(ctx context.Context, req *pb.PutVertexRequest, _ ...grpc.CallOption) (*pb.PutVertexResponse, error) {
	return connectUnary(ctx, req, a.svc.PutVertex)
}
func (a *connectClientAdapter) PutVertices(ctx context.Context, req *pb.PutVerticesRequest, _ ...grpc.CallOption) (*pb.PutVerticesResponse, error) {
	return connectUnary(ctx, req, a.svc.PutVertices)
}
func (a *connectClientAdapter) DeleteVertex(ctx context.Context, req *pb.DeleteVertexRequest, _ ...grpc.CallOption) (*pb.DeleteVertexResponse, error) {
	return connectUnary(ctx, req, a.svc.DeleteVertex)
}
func (a *connectClientAdapter) DeleteVertices(ctx context.Context, req *pb.DeleteVerticesRequest, _ ...grpc.CallOption) (*pb.DeleteVerticesResponse, error) {
	return connectUnary(ctx, req, a.svc.DeleteVertices)
}
func (a *connectClientAdapter) ScanVertices(ctx context.Context, req *pb.ScanVerticesRequest, _ ...grpc.CallOption) (*pb.ScanVerticesResponse, error) {
	return connectUnary(ctx, req, a.svc.ScanVertices)
}
func (a *connectClientAdapter) CountVerticesByPrefix(ctx context.Context, req *pb.CountVerticesByPrefixRequest, _ ...grpc.CallOption) (*pb.CountVerticesByPrefixResponse, error) {
	return connectUnary(ctx, req, a.svc.CountVerticesByPrefix)
}
func (a *connectClientAdapter) DeleteVerticesByPrefix(ctx context.Context, req *pb.DeleteVerticesByPrefixRequest, _ ...grpc.CallOption) (*pb.DeleteVerticesByPrefixResponse, error) {
	return connectUnary(ctx, req, a.svc.DeleteVerticesByPrefix)
}
func (a *connectClientAdapter) GetEdge(ctx context.Context, req *pb.GetEdgeRequest, _ ...grpc.CallOption) (*pb.GetEdgeResponse, error) {
	return connectUnary(ctx, req, a.svc.GetEdge)
}
func (a *connectClientAdapter) GetEdges(ctx context.Context, req *pb.GetEdgesRequest, _ ...grpc.CallOption) (*pb.GetEdgesResponse, error) {
	return connectUnary(ctx, req, a.svc.GetEdges)
}
func (a *connectClientAdapter) AddEdge(ctx context.Context, req *pb.AddEdgeRequest, _ ...grpc.CallOption) (*pb.AddEdgeResponse, error) {
	return connectUnary(ctx, req, a.svc.AddEdge)
}
func (a *connectClientAdapter) AddEdges(ctx context.Context, req *pb.AddEdgesRequest, _ ...grpc.CallOption) (*pb.AddEdgesResponse, error) {
	return connectUnary(ctx, req, a.svc.AddEdges)
}
func (a *connectClientAdapter) PutEdge(ctx context.Context, req *pb.PutEdgeRequest, _ ...grpc.CallOption) (*pb.PutEdgeResponse, error) {
	return connectUnary(ctx, req, a.svc.PutEdge)
}
func (a *connectClientAdapter) PutEdges(ctx context.Context, req *pb.PutEdgesRequest, _ ...grpc.CallOption) (*pb.PutEdgesResponse, error) {
	return connectUnary(ctx, req, a.svc.PutEdges)
}
func (a *connectClientAdapter) DeleteEdge(ctx context.Context, req *pb.DeleteEdgeRequest, _ ...grpc.CallOption) (*pb.DeleteEdgeResponse, error) {
	return connectUnary(ctx, req, a.svc.DeleteEdge)
}
func (a *connectClientAdapter) DeleteEdges(ctx context.Context, req *pb.DeleteEdgesRequest, _ ...grpc.CallOption) (*pb.DeleteEdgesResponse, error) {
	return connectUnary(ctx, req, a.svc.DeleteEdges)
}
func (a *connectClientAdapter) ScanEdges(ctx context.Context, req *pb.ScanEdgesRequest, _ ...grpc.CallOption) (*pb.ScanEdgesResponse, error) {
	return connectUnary(ctx, req, a.svc.ScanEdges)
}
func (a *connectClientAdapter) GetServerStatus(ctx context.Context, req *pb.GetServerStatusRequest, _ ...grpc.CallOption) (*pb.GetServerStatusResponse, error) {
	return connectUnary(ctx, req, a.svc.GetServerStatus)
}
func (a *connectClientAdapter) GetReplicationStatus(ctx context.Context, req *pb.GetReplicationStatusRequest, _ ...grpc.CallOption) (*pb.GetReplicationStatusResponse, error) {
	return connectUnary(ctx, req, a.svc.GetReplicationStatus)
}

// SubscribeConnect opens a server-stream against the replication service
// over the Connect transport and returns an iter.Seq2 yielding successive
// *SubscribeResponse frames until the stream closes. fromSeq passes
// through unchanged.
//
// Stop conditions: clean EOF, any server error, the supplied ctx being
// canceled, or the consumer returning false from yield. The underlying
// *connect.ServerStreamForClient is always closed before iteration ends.
//
// This method lives on *Lantern (not the inner adapter) because the grpc
// path historically exposes Subscribe via the pb client directly and the
// Connect path's ergonomic surface is the iterator shape; mixing both on
// the same client interface would force a transport-specific surface back
// onto every consumer of the existing grpc-backed *Lantern. The grpc-
// backed *Lantern returns an error here so misuse fails loudly rather
// than silently no-op'ing.
func (l *Lantern) SubscribeConnect(ctx context.Context, fromSeq uint64) iter.Seq2[*pb.SubscribeResponse, error] {
	return func(yield func(*pb.SubscribeResponse, error) bool) {
		if l.connectHTTPClient == nil || l.connectBaseURL == "" {
			yield(nil, errors.New("client: SubscribeConnect requires NewLanternConnect; this *Lantern was built via the grpc path"))
			return
		}
		rc := graphv1connect.NewLanternReplicationServiceClient(l.connectHTTPClient, l.connectBaseURL)
		stream, err := rc.Subscribe(ctx, connect.NewRequest(&pb.SubscribeRequest{FromSeq: fromSeq}))
		if err != nil {
			yield(nil, connectErrToGRPC(err))
			return
		}
		defer func() { _ = stream.Close() }()
		for stream.Receive() {
			if !yield(stream.Msg(), nil) {
				return
			}
		}
		if err := stream.Err(); err != nil && !errors.Is(err, io.EOF) {
			yield(nil, connectErrToGRPC(err))
		}
	}
}

// PingConnect issues a plain GET against {metricsURL}/healthz. Returns
// nil iff the server responds 200.
//
// Unlike the legacy gRPC-Health-based Ping, this probe lives entirely
// over HTTP/1.1 and does not depend on the Connect baseURL — health
// checks are owned by the metrics server (default :9090), not the
// Connect listener. Pass http://host:9090 (or wherever
// LANTERN_METRICS_ADDR points). nil httpClient defaults to
// http.DefaultClient — the metrics server speaks plain HTTP/1.1, so
// the h2c-flavored httpClient stored on *Lantern is not required.
func PingConnect(ctx context.Context, httpClient *http.Client, metricsURL string) error {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, strings.TrimRight(metricsURL, "/")+"/healthz", nil)
	if err != nil {
		return err
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("client: PingConnect %s returned %d", req.URL.String(), resp.StatusCode)
	}
	_, _ = io.Copy(io.Discard, resp.Body)
	return nil
}

File: sdks/node/src/connect-client.ts
/**
 * Connect-Node-backed Lantern client (#340 — additive, parallel to the
 * legacy grpc-js `Lantern` class).
 *
 * Why a parallel class rather than swapping `Lantern` in place: the
 * grpc-js path stays compiling and tested until the production cutover
 * (#347) and final cleanup (#342) run. This mirrors the additive
 * roll-out pattern used by every prior step in the migration (#336
 * codegen, #337 listener, #338 sdks/go, #339 admin) — main stays
 * green at every PR, and downstream consumers migrate constructor
 * call sites independently.
 *
 * The new class exposes the SAME `Vertex` / `Edge` / `Graph` value-
 * object shapes the legacy `Lantern` class returns, so consumer code
 * that holds a typed Vertex / Edge keeps compiling unchanged after
 * switching from `Lantern.connect("host:6380")` to
 * `LanternConnect.connect("http://host:6381")`.
 *
 * Built on @connectrpc/connect v2 (which renames the v1
 * `createPromiseClient` to `createClient` and the v1 `Message<T>`
 * class hierarchy to plain `Message<TypeName>` interface) and
 * @bufbuild/protobuf v2 (single-plugin codegen via
 * @bufbuild/protoc-gen-es, emitting `*Schema` schema descriptors
 * consumed by `create()` / `fromJson()` / `toJson()`).
 *
 * @packageDocumentation
 */

import { type Client, type Interceptor, createClient } from "@connectrpc/connect";
import { createConnectTransport, type ConnectTransportOptions } from "@connectrpc/connect-node";
import { create, fromJson, toJson, type JsonValue } from "@bufbuild/protobuf";

import {
  EdgeSchema,
  LanternService,
  Optimization as PbOptimization,
  VertexSchema,
} from "./gen/graph/v1/graph_pb.js";

import {
  BatchError,
  LanternError,
  InvalidArgumentError,
  NotFoundError,
  ResourceExhaustedError,
} from "./errors.js";
import { Optimization } from "./values.js";
import {
  type ConnectOptions,
  DEFAULT_BATCH_CHUNK_SIZE,
  type DeleteByPrefixOptions,
  type EdgeScanOptions,
  type IlluminateOptions,
  type ScanOptions,
} from "./options.js";
import {
  fromPbVertexJson,
  fromPbEdgeJson,
  toPbVertexJson,
  type Edge,
  type EdgeInput,
  type Graph,
  type Vertex,
  type VertexInput,
} from "./connect-values.js";

/**
 * Constructor arguments for `LanternConnect.connect`. The HTTP base
 * URL is the only required field; everything else either has a
 * sensible default or is an opt-in tuning knob.
 */
export interface LanternConnectArgs {
  options?: ConnectOptions;
  /**
   * Optional list of Connect interceptors run on every unary call.
   * The order matches @connectrpc/connect's `Interceptor[]` chain —
   * the first entry sees outgoing requests first and responses last.
   */
  interceptors?: Interceptor[];
  /**
   * Override the transport options forwarded to
   * `createConnectTransport`. `baseUrl` is filled in from the
   * constructor arg and cannot be overridden here.
   */
  transportOptions?: Omit<ConnectTransportOptions, "baseUrl">;
}

// connect-es v2 enum values match the proto numeric IDs verbatim
// — no translation needed beyond the type cast.
const OPTIMIZATION_TO_PB: Record<number, PbOptimization> = {
  [Optimization.UNSPECIFIED]: PbOptimization.UNSPECIFIED,
  [Optimization.MINIMUM_SPANNING_TREE]: PbOptimization.MINIMUM_SPANNING_TREE,
  [Optimization.MAXIMUM_SPANNING_TREE]: PbOptimization.MAXIMUM_SPANNING_TREE,
  [Optimization.SHORTEST_PATH_TREE]: PbOptimization.SHORTEST_PATH_TREE,
  [Optimization.SHORTEST_PATH_TREE_INVERSE]: PbOptimization.SHORTEST_PATH_TREE_INVERSE,
};

/**
 * Promise-based Connect-Node client for the Lantern service.
 *
 * Construct with `LanternConnect.connect("http://host:6381")` pointed
 * at the server's additive Connect listener
 * (`LANTERN_CONNECT_PORT`). All methods return Promises; pass `signal`
 * for AbortController-driven cancellation.
 *
 * Batch helpers (`putVertices`, `addEdges`, `putEdges`,
 * `deleteVertices`, `deleteEdges`) auto-chunk at
 * `ConnectOptions.batchChunkSize` (default 1000) and throw
 * `BatchError` with a resumable `written` offset on partial failure
 * — matching the legacy `Lantern` class's contract exactly.
 */
export class LanternConnect {
  private readonly client: Client<typeof LanternService>;
  private readonly options: ConnectOptions;

  private constructor(client: Client<typeof LanternService>, options: ConnectOptions) {
    this.client = client;
    this.options = options;
  }

  /**
   * Open a Connect-Node client against the supplied base URL. The
   * base URL MUST include the scheme — `http://` for h2c (matches
   * LANTERN_CONNECT_PORT default), or `https://` for TLS.
   *
   * Defaults:
   *   - HTTP/2 transport (Connect protocol, JSON codec); set
   *     `args.transportOptions.useBinaryFormat = true` to flip to
   *     protobuf.
   *   - Batch chunk size 1000 (mirror legacy Lantern).
   *   - No per-call timeout; pass
   *     `args.options.defaultTimeoutMs` to apply one.
   */
  static connect(baseUrl: string, args: LanternConnectArgs = {}): LanternConnect {
    if (!baseUrl) {
      throw new Error("LanternConnect.connect: baseUrl is required");
    }
    if (!/^https?:\/\//.test(baseUrl)) {
      throw new Error(
        `LanternConnect.connect: baseUrl must include scheme (http:// or https://); got ${JSON.stringify(baseUrl)}`,
      );
    }
    const normalised = baseUrl.replace(/\/$/, "");
    const transport = createConnectTransport({
      baseUrl: normalised,
      httpVersion: "2",
      interceptors: args.interceptors,
      ...(args.transportOptions as Partial<ConnectTransportOptions> | undefined),
    } as ConnectTransportOptions);
    return new LanternConnect(createClient(LanternService, transport), {
      batchChunkSize: DEFAULT_BATCH_CHUNK_SIZE,
      ...(args.options ?? {}),
    });
  }

  /**
   * Releases the Node http2.Session pool the Connect transport owns.
   * @connectrpc/connect-node v2 does not expose a public close hook;
   * the session pool is torn down by GC and by node's natural
   * shutdown. Kept for API parity with the legacy `Lantern.close()`.
   */
  close(): void {
    // No-op until @connectrpc/connect-node surfaces a public
    // `transport.disconnect()` method. The Node event loop closes
    // h2 sessions on process exit naturally.
  }

  // --- Vertex unary RPCs ---

  async getVertex(key: string, signal?: AbortSignal): Promise<Vertex> {
    return this.invoke(async () => {
      const resp = await this.client.getVertex({ key }, this.callOpts(signal));
      if (!resp.vertex) {
        throw new NotFoundError(`vertex not found: ${key}`);
      }
      return fromPbVertexJson(toJson(VertexSchema, resp.vertex) as Record<string, unknown>);
    });
  }

  async putVertex(input: VertexInput, signal?: AbortSignal): Promise<void> {
    const vertex = fromJson(VertexSchema, toPbVertexJson(input) as JsonValue);
    return this.invoke(async () => {
      await this.client.putVertex({ vertex }, this.callOpts(signal));
    });
  }

  async deleteVertex(key: string, signal?: AbortSignal): Promise<boolean> {
    return this.invoke(async () => {
      const resp = await this.client.deleteVertex({ key }, this.callOpts(signal));
      return resp.existed;
    });
  }

  async getVertices(
    keys: readonly string[],
    signal?: AbortSignal,
  ): Promise<{ found: Vertex[]; missing: string[] }> {
    if (keys.length === 0) return { found: [], missing: [] };
    const found: Vertex[] = [];
    const missing: string[] = [];
    await this.runBatchRead(keys, async (chunk) => {
      const resp = await this.client.getVertices({ keys: chunk }, this.callOpts(signal));
      for (const v of resp.vertices) {
        found.push(fromPbVertexJson(toJson(VertexSchema, v) as Record<string, unknown>));
      }
      for (const m of resp.missing) missing.push(m);
    });
    return { found, missing };
  }

  async putVertices(inputs: readonly VertexInput[], signal?: AbortSignal): Promise<void> {
    if (inputs.length === 0) return;
    await this.runBatchWrite(inputs, async (chunk) => {
      const vertices = chunk.map((vi) => fromJson(VertexSchema, toPbVertexJson(vi) as JsonValue));
      await this.client.putVertices({ vertices }, this.callOpts(signal));
    });
  }

  async deleteVertices(keys: readonly string[], signal?: AbortSignal): Promise<number> {
    if (keys.length === 0) return 0;
    let total = 0;
    await this.runBatchWrite(keys, async (chunk) => {
      const resp = await this.client.deleteVertices({ keys: chunk }, this.callOpts(signal));
      total += resp.deleted;
    });
    return total;
  }

  async scanVertices(
    prefix: string,
    opts: ScanOptions = {},
    signal?: AbortSignal,
  ): Promise<{ vertices: Vertex[]; nextCursor: Uint8Array }> {
    return this.invoke(async () => {
      const resp = await this.client.scanVertices(
        {
          prefix,
          limit: opts.limit ?? 0,
          cursor: opts.cursor ?? new Uint8Array(),
        },
        this.callOpts(signal),
      );
      return {
        vertices: resp.vertices.map((v) =>
          fromPbVertexJson(toJson(VertexSchema, v) as Record<string, unknown>),
        ),
        nextCursor: resp.nextCursor,
      };
    });
  }

  async *scanVerticesAll(
    prefix: string,
    batchSize?: number,
    signal?: AbortSignal,
  ): AsyncIterable<Vertex[]> {
    let cursor: Uint8Array = new Uint8Array();
    while (true) {
      if (signal?.aborted) throw new LanternError("scanVerticesAll aborted");
      const page = await this.scanVertices(prefix, { limit: batchSize ?? 0, cursor }, signal);
      yield page.vertices;
      if (page.nextCursor.length === 0) return;
      cursor = page.nextCursor;
    }
  }

  async countVerticesByPrefix(prefix: string, signal?: AbortSignal): Promise<bigint> {
    return this.invoke(async () => {
      const resp = await this.client.countVerticesByPrefix({ prefix }, this.callOpts(signal));
      return resp.count;
    });
  }

  async deleteVerticesByPrefix(
    prefix: string,
    opts: DeleteByPrefixOptions = {},
    signal?: AbortSignal,
  ): Promise<bigint> {
    return this.invoke(async () => {
      const resp = await this.client.deleteVerticesByPrefix(
        {
          prefix,
          limit: opts.limit ?? 0,
          dryRun: opts.dryRun ?? false,
        },
        this.callOpts(signal),
      );
      return resp.deleted;
    });
  }

  // --- Edge unary RPCs ---

  async getEdge(tail: string, head: string, signal?: AbortSignal): Promise<Edge> {
    return this.invoke(async () => {
      const resp = await this.client.getEdge({ tail, head }, this.callOpts(signal));
      if (!resp.edge) {
        throw new NotFoundError(`edge not found: ${tail} -> ${head}`);
      }
      return fromPbEdgeJson(toJson(EdgeSchema, resp.edge) as Record<string, unknown>);
    });
  }

  async addEdge(input: EdgeInput, signal?: AbortSignal): Promise<void> {
    const edge = fromJson(EdgeSchema, edgeInputToJson(input) as JsonValue);
    return this.invoke(async () => {
      await this.client.addEdge({ edge }, this.callOpts(signal));
    });
  }

  async putEdge(input: EdgeInput, signal?: AbortSignal): Promise<void> {
    const edge = fromJson(EdgeSchema, edgeInputToJson(input) as JsonValue);
    return this.invoke(async () => {
      await this.client.putEdge({ edge }, this.callOpts(signal));
    });
  }

  async deleteEdge(tail: string, head: string, signal?: AbortSignal): Promise<boolean> {
    return this.invoke(async () => {
      const resp = await this.client.deleteEdge({ tail, head }, this.callOpts(signal));
      return resp.existed;
    });
  }

  async getEdges(
    refs: readonly { tail: string; head: string }[],
    signal?: AbortSignal,
  ): Promise<{ found: Edge[]; missing: { tail: string; head: string }[] }> {
    if (refs.length === 0) return { found: [], missing: [] };
    const found: Edge[] = [];
    const missing: { tail: string; head: string }[] = [];
    await this.runBatchRead(refs, async (chunk) => {
      const resp = await this.client.getEdges(
        { edges: chunk.map((r) => ({ tail: r.tail, head: r.head })) },
        this.callOpts(signal),
      );
      for (const e of resp.edges) {
        found.push(fromPbEdgeJson(toJson(EdgeSchema, e) as Record<string, unknown>));
      }
      for (const m of resp.missing) {
        missing.push({ tail: m.tail, head: m.head });
      }
    });
    return { found, missing };
  }

  async addEdges(inputs: readonly EdgeInput[], signal?: AbortSignal): Promise<void> {
    if (inputs.length === 0) return;
    await this.runBatchWrite(inputs, async (chunk) => {
      const edges = chunk.map((e) => fromJson(EdgeSchema, edgeInputToJson(e) as JsonValue));
      await this.client.addEdges({ edges }, this.callOpts(signal));
    });
  }

  async putEdges(inputs: readonly EdgeInput[], signal?: AbortSignal): Promise<void> {
    if (inputs.length === 0) return;
    await this.runBatchWrite(inputs, async (chunk) => {
      const edges = chunk.map((e) => fromJson(EdgeSchema, edgeInputToJson(e) as JsonValue));
      await this.client.putEdges({ edges }, this.callOpts(signal));
    });
  }

  async deleteEdges(
    refs: readonly { tail: string; head: string }[],
    signal?: AbortSignal,
  ): Promise<number> {
    if (refs.length === 0) return 0;
    let total = 0;
    await this.runBatchWrite(refs, async (chunk) => {
      const resp = await this.client.deleteEdges(
        { edges: chunk.map((r) => ({ tail: r.tail, head: r.head })) },
        this.callOpts(signal),
      );
      total += resp.deleted;
    });
    return total;
  }

  async scanEdges(
    opts: EdgeScanOptions = {},
    signal?: AbortSignal,
  ): Promise<{ edges: Edge[]; nextCursor: Uint8Array }> {
    return this.invoke(async () => {
      const resp = await this.client.scanEdges(
        {
          tailPrefix: opts.tailPrefix ?? "",
          headPrefix: opts.headPrefix ?? "",
          limit: opts.limit ?? 0,
          cursor: opts.cursor ?? new Uint8Array(),
        },
        this.callOpts(signal),
      );
      return {
        edges: resp.edges.map((e) =>
          fromPbEdgeJson(toJson(EdgeSchema, e) as Record<string, unknown>),
        ),
        nextCursor: resp.nextCursor,
      };
    });
  }

  async *scanEdgesAll(
    opts: EdgeScanOptions = {},
    batchSize?: number,
    signal?: AbortSignal,
  ): AsyncIterable<Edge[]> {
    let cursor: Uint8Array = opts.cursor ?? new Uint8Array();
    while (true) {
      if (signal?.aborted) throw new LanternError("scanEdgesAll aborted");
      const page = await this.scanEdges(
        { ...opts, limit: batchSize ?? opts.limit ?? 0, cursor },
        signal,
      );
      yield page.edges;
      if (page.nextCursor.length === 0) return;
      cursor = page.nextCursor;
    }
  }

  async illuminate(
    seed: string,
    opts: IlluminateOptions = {},
    signal?: AbortSignal,
  ): Promise<Graph> {
    if (!seed) throw new InvalidArgumentError("illuminate: seed is required");
    return this.invoke(async () => {
      const resp = await this.client.illuminate(
        {
          seed,
          step: opts.step ?? 0,
          k: opts.k ?? 0,
          tfidf: opts.tfidf ?? false,
          optimization:
            OPTIMIZATION_TO_PB[opts.optimization ?? Optimization.UNSPECIFIED] ??
            PbOptimization.UNSPECIFIED,
        },
        this.callOpts(signal),
      );
      const vertices = new Map<string, Vertex>();
      for (const v of resp.graph?.vertices ?? []) {
        const flat = toJson(VertexSchema, v) as Record<string, unknown>;
        vertices.set(String(flat.key ?? ""), fromPbVertexJson(flat));
      }
      const edges = new Map<string, Map<string, number>>();
      for (const e of resp.graph?.edges ?? []) {
        const flat = toJson(EdgeSchema, e) as Record<string, unknown>;
        const tail = String(flat.tail ?? "");
        const head = String(flat.head ?? "");
        const weight = Number(flat.weight ?? 0);
        if (!edges.has(tail)) edges.set(tail, new Map());
        edges.get(tail)!.set(head, weight);
      }
      return { vertices, edges };
    });
  }

  // --- internals ---

  private callOpts(signal?: AbortSignal): {
    signal?: AbortSignal;
    timeoutMs?: number;
  } {
    const out: { signal?: AbortSignal; timeoutMs?: number } = {};
    if (signal) out.signal = signal;
    if (this.options.defaultTimeoutMs && this.options.defaultTimeoutMs > 0) {
      out.timeoutMs = this.options.defaultTimeoutMs;
    }
    return out;
  }

  private async invoke<T>(fn: () => Promise<T>): Promise<T> {
    try {
      return await fn();
    } catch (err) {
      throw wrapConnectError(err);
    }
  }

  private chunkSize(): number {
    const n = this.options.batchChunkSize ?? DEFAULT_BATCH_CHUNK_SIZE;
    return n > 0 ? n : DEFAULT_BATCH_CHUNK_SIZE;
  }

  private async runBatchWrite<T>(
    items: readonly T[],
    sendChunk: (chunk: T[]) => Promise<void>,
  ): Promise<void> {
    const size = this.chunkSize();
    let written = 0;
    for (let i = 0; i < items.length; i += size) {
      const chunk = items.slice(i, Math.min(i + size, items.length));
      try {
        await sendChunk(chunk);
      } catch (err) {
        throw new BatchError(written, wrapConnectError(err));
      }
      written += chunk.length;
    }
  }

  private async runBatchRead<T>(
    items: readonly T[],
    sendChunk: (chunk: T[]) => Promise<void>,
  ): Promise<void> {
    const size = this.chunkSize();
    for (let i = 0; i < items.length; i += size) {
      const chunk = items.slice(i, Math.min(i + size, items.length));
      try {
        await sendChunk(chunk);
      } catch (err) {
        throw wrapConnectError(err);
      }
    }
  }
}

function edgeInputToJson(input: EdgeInput): Record<string, unknown> {
  const out: Record<string, unknown> = {
    tail: input.tail,
    head: input.head,
    weight: input.weight,
  };
  const exp = resolveExpiration(input.ttlSeconds, input.expiration);
  if (exp) out.expiration = exp.toISOString();
  return out;
}

function resolveExpiration(ttlSeconds?: number, expiration?: Date): Date | undefined {
  if (expiration !== undefined && ttlSeconds !== undefined) {
    throw new TypeError("specify either ttlSeconds or expiration, not both");
  }
  if (expiration !== undefined) return expiration;
  if (ttlSeconds !== undefined) return new Date(Date.now() + ttlSeconds * 1000);
  return undefined;
}

/**
 * Translates a thrown Connect error into the typed LanternError
 * hierarchy. Non-Connect errors fall through unchanged so cancellation
 * (AbortError) and network failures keep their native shape.
 */
function wrapConnectError(err: unknown): LanternError {
  if (err instanceof LanternError) return err;
  // ConnectError exposes a numeric `.code` matching the gRPC code set.
  // Duck-type to avoid pulling in the ConnectError type.
  const ce = err as { code?: number; rawMessage?: string; message?: string };
  const message = ce.rawMessage ?? ce.message ?? String(err);
  // Code numbers (mirror @connectrpc/connect Code enum):
  // 5 = NotFound, 3 = InvalidArgument, 8 = ResourceExhausted.
  switch (ce.code) {
    case 5:
      return new NotFoundError(message, { cause: err });
    case 3:
      return new InvalidArgumentError(message, { cause: err });
    case 8:
      return new ResourceExhaustedError(message, { cause: err });
    default:
      return new LanternError(message, { cause: err });
  }
}

// Suppress the "unused" warning on create — it is part of the
// protobuf-es v2 public API surface and may be needed by future
// edits inside this file.
void create;
